src/ServiceControl.Monitoring/EndpointFactory.cs

```diff
@@ -76,12 +82,32 @@ public static void MakeMetricsReceiver(EndpointConfiguration config, Settings se
             recoverability.DisableLegacyRetriesSatellite();
         }
 
-        static IContainer CreateContainer(Settings settings)
+        static Func<QueueLengthStore, IProvideQueueLength> QueueLengthProviderBuilder(string explicitConnectionStringValue, Type transportType)
+        {
+            return qls =>
+            {
+                var connectionString = explicitConnectionStringValue ?? ConfigurationManager.ConnectionStrings["NServiceBus/Transport"]?.ConnectionString;
+
+                var queueLengthProviderType = transportType.Assembly.GetTypes()
+                    .SingleOrDefault(p => typeof(IProvideQueueLength).IsAssignableFrom(p));
```

Comment:
Just a thought: do want to fail on multiple implementations or should we pick up one and just log it? With the current approach, `SingleOrDefault` will throw on multiple ones.

Reply:
I can't think of scenario in which transport would have more than one implementation. I would go with fail fast approach and keep `SingelOrDefault`.